server/src/main/java/io/druid/server/coordinator/CachingCostBalancerStrategyFactory.java

```diff
@@ -0,0 +1,159 @@
+import io.druid.client.ServerView;
+import io.druid.concurrent.Execs;
+import io.druid.guice.ManageLifecycle;
+import io.druid.java.util.common.ISE;
+import io.druid.java.util.common.lifecycle.LifecycleStart;
+import io.druid.java.util.common.lifecycle.LifecycleStop;
+import io.druid.server.coordination.DruidServerMetadata;
+import io.druid.server.coordinator.cost.ClusterCostCache;
+import io.druid.timeline.DataSegment;
+
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+@ManageLifecycle
+public class CachingCostBalancerStrategyFactory implements BalancerStrategyFactory
+{
+  private static final EmittingLogger log = new EmittingLogger(CachingCostBalancerStrategyFactory.class);
+
+  private final ServerInventoryView serverInventoryView;
+  private final AtomicBoolean started = new AtomicBoolean(false);
+  private final AtomicReference<ExecutorService> executorRef = new AtomicReference<>(null);
+  private volatile boolean initialized = false;
+  private ClusterCostCache.Builder clusterCostCacheBuilder;
+
+  @Inject
+  public CachingCostBalancerStrategyFactory(ServerInventoryView serverInventoryView)
+  {
+    this.serverInventoryView = Preconditions.checkNotNull(serverInventoryView);
+  }
+
+  @LifecycleStart
+  public void start()
+  {
+    if (!executorRef.compareAndSet(null, Execs.singleThreaded("CachingCostBalancerStrategy-executor"))) {
+      throw new ISE("CachingCostBalancerStrategyFactory is already started or has been stopped");
+    }
+    if (!started.compareAndSet(false, true)) {
+      throw new ISE("CachingCostBalancerStrategyFactory is already started");
+    }
+    serverInventoryView.registerSegmentCallback(
+        executorRef.get(),
+        new ServerView.SegmentCallback()
+        {
+          @Override
+          public ServerView.CallbackAction segmentAdded(
+              DruidServerMetadata server, DataSegment segment
+          )
+          {
+            clusterCostCacheBuilder.addSegment(server.getName(), segment);
+            return ServerView.CallbackAction.CONTINUE;
+          }
+
+          @Override
+          public ServerView.CallbackAction segmentRemoved(
+              DruidServerMetadata server, DataSegment segment
+          )
+          {
+            clusterCostCacheBuilder.removeSegment(server.getName(), segment);
+            return ServerView.CallbackAction.CONTINUE;
+          }
+
+          @Override
+          public ServerView.CallbackAction segmentViewInitialized()
+          {
+            initialized = true;
+            clusterCostCacheBuilder = ClusterCostCache.builder();
```

Comment:
This line doesn't solve the problem, because you always construct `CachingCostBalancerStrategy` from an empty builder.